rustls/src/common_state.rs

```diff
@@ -845,6 +855,30 @@ pub(crate) trait State<Data>: Send + Sync {
     fn into_owned(self: Box<Self>) -> Box<dyn State<Data> + 'static>;
 }
 
+/// This is an output of [`State::handle()`]
+///
+/// It contains the next step in the state machine.
+pub(crate) enum Transition<'a, Data> {
+    /// Transition to the next state.
+    ///
+    /// The current state was consumed.
+    Next(Box<dyn State<Data> + 'a>),
+
+    /// The current state is blocked.
+    ///
+    /// `current` is the current state; returned so it is not consumed.
+    /// Likewise, `message` is the current message.
+    /// `why` describes the operation that it is waiting on.
+    ///
+    /// `handle()` should be called again with the same message.  It will
+    /// make progress when possible.
```

Comment:
Perhaps this is a place to discuss the more involved decision that could be done based on `why` to decide when to call `handle()` again?